Importing basic libraries for data processing

In [2]:
import numpy as np
import pandas as pd
from sklearn import metrics

Loading the dataset

In [3]:
data = pd.read_csv('./Dataset/Existing Base.csv')
data.head()

REF_NO children age_band                status        occupation  \
0       1     Zero    51-55               Partner     Manual Worker   
1       2     Zero    55-60  Single/Never Married           Retired   
2       3     Zero    26-30  Single/Never Married      Professional   
3       5     Zero    18-21  Single/Never Married      Professional   
4       6     Zero    45-50               Partner  Business Manager   

  occupation_partner home_status      family_income self_employed  \
0  Secretarial/Admin    Own Home  <17,500, >=15,000            No   
1            Retired    Own Home  <27,500, >=25,000            No   
2              Other    Own Home  <30,000, >=27,500           Yes   
3      Manual Worker    Own Home  <15,000, >=12,500            No   
4            Unknown    Own Home  <30,000, >=27,500            No   

  self_employed_partner  ...  Investment Tax Saving Bond Home Loan  \
0                    No  ...                       19.99      0.00   
1                    No  ...                        0.00      0.00   
2                    No  ...                        0.00      3.49   
3                    No  ...                        0.00      0.00   
4                    No  ...                        0.00     45.91   

  Online Purchase Amount Revenue Grid  gender         region  \
0                   0.00            1  Female          Wales   
1                   0.00            2  Female     North West   
2                   0.00            2    Male          North   
3                   0.00            2  Female  West Midlands   
4                  25.98            2  Female       Scotland   

   Investment in Commudity  Investment in Equity  Investment in Derivative  \
0                    74.67                 18.66                     32.32   
1                    20.19                  0.00                      4.33   
2                    98.06                 31.07                     80.96   
3                     4.10                 14.15                     17.57   
4                    70.16                 55.86                     80.44   

   Portfolio Balance  
0              89.43  
1              22.78  
2             171.78  
3             -41.70  
4             235.02  

[5 rows x 32 columns]

In [4]:
data.shape

(10155, 32)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10155 entries, 0 to 10154
Data columns (total 32 columns):
REF_NO                             10155 non-null int64
children                           10155 non-null object
age_band                           10155 non-null object
status                             10155 non-null object
occupation                         10155 non-null object
occupation_partner                 10155 non-null object
home_status                        10155 non-null object
family_income                      10155 non-null object
self_employed                      10155 non-null object
self_employed_partner              10155 non-null object
year_last_moved                    10155 non-null int64
TVarea                             10155 non-null object
post_code                          10155 non-null object
post_area                          10155 non-null object
Average Credit Card Transaction    10155 non-null float64
Balance Transfer                   10

In [6]:
len(data.select_dtypes(['object']).columns)

14

We are having 32 columns out of them 14 are categorical. Lets look each variable one by one

First we are having Ref.No we can drop that variable so keep that aside as of now and lets look into 2nd feature which is *children*.

In [7]:
data['children'].value_counts()

Zero    6208
1       1848
2       1607
3        473
4+        19
Name: children, dtype: int64

In [8]:
data.loc[data['children'] == 'Zero','children'] = '0'
data.loc[data['children'] == '4+','children'] = '4'
data['children'] = pd.to_numeric(data['children'])
data['children'].dtype

dtype('int64')

In [9]:
data['children'].value_counts()

0    6208
1    1848
2    1607
3     473
4      19
Name: children, dtype: int64

Our Target variable is Revenue.Grid so converting value 2 to 0 and 1 to 1

In [10]:
data['Revenue Grid'].value_counts()

2    9069
1    1086
Name: Revenue Grid, dtype: int64

In [11]:
data['Revenue Grid'] = np.where(data['Revenue Grid'] == 2,0,1)
data['Revenue Grid'].value_counts

<bound method IndexOpsMixin.value_counts of 0        1
1        0
2        0
3        0
4        0
        ..
10150    0
10151    0
10152    0
10153    0
10154    0
Name: Revenue Grid, Length: 10155, dtype: int64>

Now lets move to age_band variable

In [12]:
data['age_band'].value_counts()

45-50      1359
36-40      1134
41-45      1112
31-35      1061
51-55      1052
55-60      1047
26-30       927
61-65       881
65-70       598
22-25       456
71+         410
18-21        63
Unknown      55
Name: age_band, dtype: int64

considering the age as a categorical variable, so instead of taking mean value just creating dummies based on the response on target variable

In [13]:
round(data.groupby('age_band')['Revenue Grid'].mean(),2)

age_band
18-21      0.17
22-25      0.11
26-30      0.11
31-35      0.11
36-40      0.13
41-45      0.11
45-50      0.10
51-55      0.10
55-60      0.11
61-65      0.09
65-70      0.10
71+        0.10
Unknown    0.05
Name: Revenue Grid, dtype: float64

In [14]:
for i in range(len(data)):
    if data["age_band"][i] in ["71+","65-70","51-55","45-50"]:
        data.loc[i,"age_band"]="ab_10"
    if data["age_band"][i] in ["55-60","41-45","31-35","22-25","26-30"]:
        data.loc[i,"age_band"]="ab_11"
    if data["age_band"][i]=="36-40":
        data.loc[i,"age_band"]="ab_13"
    if data["age_band"][i]=="18-21":
        data.loc[i,"age_band"]="ab_17"
    if data["age_band"][i]=="61-65":
        data.loc[i,"age_band"]="ab_9"
ab_dummies=pd.get_dummies(data["age_band"])
ab_dummies.head()

Unknown  ab_10  ab_11  ab_13  ab_17  ab_9
0        0      1      0      0      0     0
1        0      0      1      0      0     0
2        0      0      1      0      0     0
3        0      0      0      0      1     0
4        0      1      0      0      0     0

In [15]:
ab_dummies.drop('Unknown',axis='columns',inplace=True)

In [16]:
data = pd.concat((data,ab_dummies),axis='columns')

In [17]:
data.drop('age_band',axis='columns',inplace=True)

We have just created dummies age_band by grouping them. Now go for status feature

In [18]:
data['status'].value_counts()

Partner                 7709
Single/Never Married    1101
Divorced/Separated       679
Widowed                  618
Unknown                   48
Name: status, dtype: int64

For this status feature also lets create dummies by not creating dummies for widowed and unknown

In [19]:
data['status_partner'] = np.where(data['status'] == 'Partner',1,0)
data['statuss_S/NM'] = np.where(data['status'] == 'Single/Never Married',1,0)
data['status_div'] = np.where(data['status'] == 'Divorced/Separated',1,0)
data.drop('status',axis='columns')

REF_NO  children         occupation occupation_partner  \
0           1         0      Manual Worker  Secretarial/Admin   
1           2         0            Retired            Retired   
2           3         0       Professional              Other   
3           5         0       Professional      Manual Worker   
4           6         0   Business Manager            Unknown   
...       ...       ...                ...                ...   
10150   11512         1  Secretarial/Admin      Manual Worker   
10151   11513         0      Manual Worker      Manual Worker   
10152   11514         2          Housewife       Professional   
10153   11516         0              Other      Manual Worker   
10154   11518         1            Unknown          Housewife   

                home_status      family_income self_employed  \
0                  Own Home  <17,500, >=15,000            No   
1                  Own Home  <27,500, >=25,000            No   
2                  Own Home  <30,000, >=27,500           Yes   
3                  Own Home  <15,000, >=12,500            No   
4                  Own Home  <30,000, >=27,500            No   
...                     ...                ...           ...   
10150              Own Home  <30,000, >=27,500            No   
10151  Rent from Council/HA  <25,000, >=22,500            No   
10152              Own Home           >=35,000            No   
10153              Own Home  <10,000, >= 8,000            No   
10154              Own Home  <27,500, >=25,000            No   

      self_employed_partner  year_last_moved       TVarea  ...  \
0                        No             1972          HTV  ...   
1                        No             1998      Granada  ...   
2                        No             1996    Tyne Tees  ...   
3                        No             1997      Central  ...   
4                        No             1995  Scottish TV  ...   
...                     ...              ...          ...  ...   
10150                    No             1972      Carlton  ...   
10151                    No             1988     Meridian  ...   
10152                    No             1992      Central  ...   
10153                   Yes             1970      Carlton  ...   
10154                    No             1996          HTV  ...   

      Investment in Derivative Portfolio Balance  ab_10  ab_11  ab_13  ab_17  \
0                        32.32             89.43      1      0      0      0   
1                         4.33             22.78      0      1      0      0   
2                        80.96            171.78      0      1      0      0   
3                        17.57            -41.70      0      0      0      1   
4                        80.44            235.02      1      0      0      0   
...                        ...               ...    ...    ...    ...    ...   
10150                     9.57             10.23      1      0      0      0   
10151                    36.40            102.62      1      0      0      0   
10152                    14.07             76.18      0      1      0      0   
10153                     1.66              4.79      1      0      0      0   
10154                    99.35            227.12      0      1      0      0   

       ab_9  status_partner  statuss_S/NM  status_div  
0         0               1             0           0  
1         0               0             1           0  
2         0               0             1           0  
3         0               0             1           0  
4         0               1             0           0  
...     ...             ...           ...         ...  
10150     0               1             0           0  
10151     0               1             0           0  
10152     0               1             0           0  
10153     0               1             0           0  
10154     0               1             0           0  

[10155 rows x 38 columns]

For occupation and occupation partner we also create dummies by group with the response rate

In [20]:
data['occupation'].value_counts()

Professional         2449
Retired              2206
Secretarial/Admin    1797
Housewife            1255
Business Manager      732
Unknown               567
Manual Worker         556
Other                 537
Student                56
Name: occupation, dtype: int64

In [21]:
round(data.groupby('occupation')['Revenue Grid'].mean(),2)

occupation
Business Manager     0.12
Housewife            0.09
Manual Worker        0.11
Other                0.11
Professional         0.12
Retired              0.10
Secretarial/Admin    0.11
Student              0.11
Unknown              0.11
Name: Revenue Grid, dtype: float64

In [22]:
for i in range(len(data)):
    if data['occupation'][i] in ['Business Manager','Professional']:
        data.loc[i,'occupation'] = 'occ_12'
    if data['occupation'][i] in ['Manual Worker','Other','Secretarial/Admin','Student','Unknown']:
        data.loc[i,'occupation'] = 'occ_11'
    if data['occupation'][i] in ['Retired']:
        data.loc[i,'occupation'] = 'occ_9'

d = pd.get_dummies(data['occupation'])
d

Housewife  occ_11  occ_12  occ_9
0              0       1       0      0
1              0       0       0      1
2              0       0       1      0
3              0       0       1      0
4              0       0       1      0
...          ...     ...     ...    ...
10150          0       1       0      0
10151          0       1       0      0
10152          1       0       0      0
10153          0       1       0      0
10154          0       1       0      0

[10155 rows x 4 columns]

In [23]:
data = pd.concat([data,d],axis='columns')

In [24]:
data.drop(['Housewife','REF_NO','status'],axis='columns',inplace=True)

In [25]:
data.drop('occupation',axis='columns',inplace=True)

In [26]:
data['occupation_partner'].value_counts()

Unknown              2394
Professional         2051
Retired              1936
Manual Worker        1508
Business Manager      737
Secretarial/Admin     662
Housewife             527
Other                 324
Student                16
Name: occupation_partner, dtype: int64

In [27]:
round(data.groupby('occupation_partner')['Revenue Grid'].mean(),2)

occupation_partner
Business Manager     0.11
Housewife            0.11
Manual Worker        0.11
Other                0.10
Professional         0.11
Retired              0.10
Secretarial/Admin    0.12
Student              0.12
Unknown              0.10
Name: Revenue Grid, dtype: float64

In [28]:
for i in range(len(data)):
    if data['occupation_partner'][i] in ['Business Manager','Housewife','Manual Worker','Professional']:
        data.loc[i,'occupation_partner'] = 'ooc_ptr_11'
    if data.loc[i,'occupation_partner'] in ['Secretarial/Admin','Student']:
        data.loc[i,'occupation_partner'] = 'ooc_ptr_12'
d = pd.get_dummies(data['occupation_partner'])
d

Other  Retired  Unknown  ooc_ptr_11  ooc_ptr_12
0          0        0        0           0           1
1          0        1        0           0           0
2          1        0        0           0           0
3          0        0        0           1           0
4          0        0        1           0           0
...      ...      ...      ...         ...         ...
10150      0        0        0           1           0
10151      0        0        0           1           0
10152      0        0        0           1           0
10153      0        0        0           1           0
10154      0        0        0           1           0

[10155 rows x 5 columns]

In [29]:
data = pd.concat((data,d),axis=1)

In [30]:
data.drop(['Other','Retired','Unknown','occupation_partner'],axis=1,inplace=True)

Converting self_employed,self_employed_partner  to numeric

In [31]:
data['self_employed'] = np.where(data['self_employed'] == 'Yes',1,0)
data['self_employed_partner'] = np.where(data['self_employed_partner'] == 'Yes',1,0)

Now lets look into home_status

In [32]:
data['home_status'].value_counts()

Own Home                9413
Rent from Council/HA     322
Rent Privately           261
Live in Parental Hom     109
Unclassified              50
Name: home_status, dtype: int64

Own home are widly spread so we can drop the rest of the categoru and make own home as 1

In [33]:
data['home_status'] = np.where(data['home_status'] =='Own Home',1,0)

In [34]:
data['TVarea'].value_counts()

Central          1618
Carlton          1541
Meridian         1226
Yorkshire        1042
Granada          1039
HTV               845
Anglia            751
Tyne Tees         536
Scottish TV       520
TV South West     363
Grampian          227
Unknown           189
Ulster            159
Border             99
Name: TVarea, dtype: int64

In [35]:
data['post_area'].nunique()

2039

In [36]:
data['post_code'].nunique()

10040

In [37]:
data['region'].value_counts()

South East          2100
North West          1896
Unknown             1080
South West           971
West Midlands        820
Scotland             787
East Midlands        771
North                571
Wales                556
East Anglia          423
Northern Ireland     159
Isle of Man           16
Channel Islands        5
Name: region, dtype: int64

dropping TVarea, postcode,postarea,region. if model is not well performing we'll generate dummies for TVarea and region

In [38]:
data.drop(['post_area','post_code','region','TVarea'],axis='columns',inplace=True)

Now lets see gender variable

In [39]:
data['gender'].value_counts()

Female     7634
Male       2486
Unknown      35
Name: gender, dtype: int64

In [40]:
data['gender'] = np.where(data['gender'] == 'Female',0,1)

Now we left out with family income feature.

In [41]:
data['family_income'].value_counts()

>=35,000             2517
<27,500, >=25,000    1227
<30,000, >=27,500     994
<25,000, >=22,500     833
<20,000, >=17,500     683
<12,500, >=10,000     677
<17,500, >=15,000     634
<15,000, >=12,500     629
<22,500, >=20,000     590
<10,000, >= 8,000     563
< 8,000, >= 4,000     402
< 4,000               278
Unknown               128
Name: family_income, dtype: int64

In [42]:
data["fi"]=4 # by doing this , we have essentially clubbed <4000 and Unknown values . How?
data.loc[data["family_income"]=="< 8,000, >= 4,000","fi"]=6
data.loc[data["family_income"]=="<10,000, >= 8,000","fi"]=9
data.loc[data["family_income"]=="<12,500, >=10,000","fi"]=11.25
data.loc[data["family_income"]=="<15,000, >=12,500","fi"]=13.75
data.loc[data["family_income"]=="<17,500, >=15,000","fi"]=16.25
data.loc[data["family_income"]=="<20,000, >=17,500","fi"]=18.75
data.loc[data["family_income"]=="<22,500, >=20,000","fi"]=21.25
data.loc[data["family_income"]=="<25,000, >=22,500","fi"]=23.75
data.loc[data["family_income"]=="<27,500, >=25,000","fi"]=26.25
data.loc[data["family_income"]=="<30,000, >=27,500","fi"]=28.75
data.loc[data["family_income"]==">=35,000","fi"]=35
data=data.drop(["family_income"],axis=1)

In [43]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10155 entries, 0 to 10154
Data columns (total 36 columns):
children                           10155 non-null int64
home_status                        10155 non-null int64
self_employed                      10155 non-null int64
self_employed_partner              10155 non-null int64
year_last_moved                    10155 non-null int64
Average Credit Card Transaction    10155 non-null float64
Balance Transfer                   10155 non-null float64
Term Deposit                       10155 non-null float64
Life Insurance                     10155 non-null float64
Medical Insurance                  10155 non-null float64
Average A/C Balance                10155 non-null float64
Personal Loan                      10155 non-null float64
Investment in Mutual Fund          10155 non-null float64
Investment Tax Saving Bond         10155 non-null float64
Home Loan                          10155 non-null float64
Online Purchase Amount         

In [44]:
data.dropna(inplace=True)

## Splitting the data into train and test

In [45]:
from sklearn.model_selection import train_test_split

In [46]:
train,test = train_test_split(data,test_size=0.2,random_state=4)

In [47]:
x_train = train.drop('Revenue Grid',axis=1)
y_train = train['Revenue Grid']

x_test = test.drop('Revenue Grid',axis=1)
y_test = test['Revenue Grid']

### Performance metrics for Classification model

In [48]:
class Classification_Performance:
    '''Logistic_Performance(actual=<true values>,predicted=<model outcomes>)
        method -> accuracy() 
                    return accuracy of logistic model
        method -> confusion_matrix()
                    return TN,FP,FN,TP
        method -> precision_recall
                    return precision and recall of the model
        method -> ks_score()
        
        method -> f1_score()
        
        method -> fBeta_score()
        
        method -> roc_auc()
                    '''
    def __init__(self,actual,predicted):
        self.actual = actual
        self.predicted = predicted
    
    def accuracy(self):
        self.acc = metrics.accuracy_score(self.actual,self.predicted)
        print('Accuracy : ',self.acc)
        
    def confusion_matrix(self):
        cm = metrics.confusion_matrix(self.actual,self.predicted,)
        self.TN,self.FP,self.FN,self.TP = cm.ravel()
        print('True Negative : ',self.TN)
        print('False Positive : ',self.FP)
        print('False Negative : ',self.FN)
        print('True Positive : ',self.TP)
        
    def precision_recall(self):
        self.precision = self.TP/(self.TP + self.FP)
        self.recall = self.TP/(self.TP + self.FN)
        
        print("Precision : ",self.precision)
        print("Recall : ",self.recall)
        
        
    def ks_score(self):
        self.KS_score = self.recall - self.FP/(self.FP + self.TN)
        print("KS_score : ",self.KS_score)
        
    def f1_score(self):
        self.f1 = metrics.f1_score(self.actual,self.predicted)
        print("F1 Score : ",self.f1)
        
    def fBeta_score(self,Beta):
        """
        The `beta` parameter determines the weight of recall in the combined
score. ``beta < 1`` lends more weight to precision, while ``beta > 1``
favors recall (``beta -> 0`` considers only precision, ``beta -> inf``
only recall).
        """
        self.fBeta = metrics.fbeta_score(self.actual,self.predicted,beta=Beta)
        print("F Beta Score when Beta = ",Beta,"is = ",self.fBeta)
        
    def roc_auc(self):
        self.auc_roc = metrics.roc_auc_score(self.actual,self.predicted)
        print("Area under the curve = ",self.auc_roc)
        
    def log_loss(self,predicted_proba):
        self.logLoss = metrics.log_loss(self.actual,predicted_proba)
        print("My model is lacking the confidence of ",round(self.logLoss*100,2), "%")
        
    def check_all(self):
        self.accuracy()
        self.confusion_matrix()
        self.precision_recall()
        self.ks_score()
        self.f1_score()
        self.roc_auc()

# XGBoost

In [49]:
from sklearn.model_selection import RandomizedSearchCV
from xgboost import XGBClassifier

In [50]:
cl = XGBClassifier(n_jobs=-1,random_state=2)

In [51]:
params = {
'max_depth' : [2,3,4,5,6,7,8,9,10,11],
'learning_rate' : [0.01,0.05,0.09,0.1,0.15,0.2,0.25],
'n_estimators' : [100,200,300,400,500,600,700,800,900,1000,1200],
'subsample' : [0.5,0.7,0.8,0.9,1.0],
'colsample_bytree' : [0.5,0.6,0.7,0.8,0.9,1.0]
}

In [52]:
iterations = 1000
random_cl = RandomizedSearchCV(cl,param_distributions=params,n_jobs=-1,n_iter=iterations,
                              cv=3,scoring='roc_auc',random_state=2)

In [53]:
random_cl.fit(x_train,y_train)

RandomizedSearchCV(cv=3, error_score='raise-deprecating',
                   estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                           colsample_bylevel=1,
                                           colsample_bynode=1,
                                           colsample_bytree=1, gamma=0,
                                           learning_rate=0.1, max_delta_step=0,
                                           max_depth=3, min_child_weight=1,
                                           missing=None, n_estimators=100,
                                           n_jobs=-1, nthread=None,
                                           objective='binary:logistic',
                                           random_state=2, reg_alpha=...
                   param_distributions={'colsample_bytree': [0.5, 0.6, 0.7, 0.8,
                                                             0.9, 1.0],
                                        'learning_rate': [0.01, 0.05,

In [60]:
random_cl.best_score_

0.9907408334537278

In [61]:
random_cl.best_params_

{'subsample': 0.8,
 'n_estimators': 800,
 'max_depth': 6,
 'learning_rate': 0.05,
 'colsample_bytree': 1.0}

In [62]:
random_cl.best_estimator_

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1.0, gamma=0,
              learning_rate=0.05, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=None, n_estimators=800, n_jobs=-1,
              nthread=None, objective='binary:logistic', random_state=2,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=0.8, verbosity=1)

In [63]:
my_cl = random_cl.best_estimator_

In [64]:
my_cl.fit(x_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1.0, gamma=0,
              learning_rate=0.05, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=None, n_estimators=800, n_jobs=-1,
              nthread=None, objective='binary:logistic', random_state=2,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=0.8, verbosity=1)

In [65]:
pred_inClass = my_cl.predict(x_test)
pred_proba = my_cl.predict_proba(x_test)
performance = Classification_Performance(y_test,pred_inClass)
performance.check_all()
performance.log_loss(pred_proba)

Accuracy :  0.9748892171344166
True Negative :  1798
False Positive :  14
False Negative :  37
True Positive :  182
Precision :  0.9285714285714286
Recall :  0.8310502283105022
KS_score :  0.8233239589948289
F1 Score :  0.8771084337349399
Area under the curve =  0.9116619794974145
My model is lacking the confidence of  6.46 %


## Over all conclusion

**Logistic Regression**

Accuracy : 0.914327917282127

Area under the curve = 0.9018075841422479

F1 Score : 0.6903914590747331

My model is lacking the confidence of 25.86 %

**Decision Tree with out CV and Max_depth = 7**

Accuracy : 0.9290989660265879

Area under the curve = 0.9040642293386555

F1 Score : 0.726235741444867 (Improved)

My model is lacking the confidence of 32.55 % 

**Decision Tree with Hyper parameter-Max_Depth tuning and withourt CV**

Accuracy :  0.9384539635647464

Area under the curve =  0.9193428891106474

F1 Score :  0.758220502901354

My model is lacking the confidence of  48.54 %

**Decision Tree with CV and Hyper parameter-Max_Depth Tuning**

Accuracy :  0.9015263417035942

F1 Score :  0.6721311475409837

***Area under the curve =  0.9167120263691071***

*My model is lacking the confidence of  29.22 %*

**Random forest with only ntrees and max depth**

Accuracy :  0.9650418513047759

F1 Score :  0.8220551378446115

Area under the curve =  0.8700142127067646

My model is lacking the confidence of  8.35 %

**Random forest with few parameters hyper tuning 200 iterations**

Accuracy :  0.965534219596258

F1 Score :  0.825

Area under the curve =  0.8722973177295956

My model is lacking the confidence of  8.35 %

**Extra Trees with only ntrees and max depth**

Accuracy :  0.96602658788774

F1 Score :  0.823529411764706

Area under the curve =  0.8645445885874989

My model is lacking the confidence of  8.89 %

**Extra Trees with few parameters hyper tuning 200 iterations**

Accuracy :  0.9630723781388478

F1 Score :  0.8329621380846325

Area under the curve =  0.9150752971060511

My model is lacking the confidence of  11.99 %

**Gradient Boosting**

Accuracy :  0.9724273756770064

F1 Score :  0.8653846153846154

Area under the curve =  0.906267954882216

My model is lacking the confidence of  95.23 %

**XGBoost**

Accuracy :  0.9748892171344166
 
F1 Score :  0.8771084337349399

Area under the curve =  0.9116619794974145

My model is lacking the confidence of  6.46 %